### ISSUE TITLE
Collada shape model limitations

### ISSUE THREAD

**Issue by angelozinzi** [NASAWorldWind/WebWorldWind#631]
Hi, I am trying to figure out if World Wind Web could be used as a novel visualization system for MATISSE (http://openplanetary.co/blog/tools/matisse.html).
Even if it should be easily adapted for larger objects (such planets and moons), for small, irregular objects (comets, asteroid) a 3D shape model needs to be used.
I therefore tried to use a shape model in collada format for the comet 67P (target of the ESA Rosetta mission) and I found out that large shape models (i.e., larger than 100000 facets) are not correctly visualized (only in World Wind Web, whereas they are well displayed with normal collada viewers).
To be precise, the object is visualized up to facet number 100000 (more or less, I don't know exactly the value, but I noted differences when using files with more than 100000 facets).

Is there a limiting keyword somewhere in the World Wind Web code? Could this be easily disabled if not needed?

Furthermore, if I try to put the shape model in the centre of the reference system (the origin, the center of the Earth) the shape disappears. I can use every altitude (negative too) larger than -6378137. With this exact value I have no object displayed.

Even in this case I would like to know if it is possibile to easily overcome this problem or it is reccomended not to use World Wind Web to visualize such objects.

Thanks,
Angelo

**Comment by strikerM** [comment]
Hi,

Can you provide a minimal test case with the large model?

Objects at the center of the Earth are probably culled by the gpu.
Try to use a bigger altitude, for example -6000000 or -5000000

**Comment by angelozinzi** [comment]
Hi, in attachment 3 test cases:
 - 67p_lo is a shape model with 31456 vertices and 62908 polygons
 - 67p_mid has 833538 vertices and 1667072 polygons
 - pumpkin has 144016 vertices and 288028 polygons


![67p_lo](https://user-images.githubusercontent.com/14196558/38725413-8d94841a-3f07-11e8-9773-30748b2de399.png)
![67p_mid](https://user-images.githubusercontent.com/14196558/38725414-8db02210-3f07-11e8-8cf5-ba76b8ab1550.png)
![pumpkin](https://user-images.githubusercontent.com/14196558/38725415-8dcbf602-3f07-11e8-9370-0dd2446ec405.png)

As you can see, 67p_lo is correctly visualized (on the left the world wind web visualization, on the right the collada visualization with a python viewer); the pumpkin shape is visualized without the petiole from the body; 67p_mid is very poorly visualized (I think only a small part of a lobe of the shape is shown).

This is indicative of a limit around 100000 (vertices, maybe?).

Sorry for not providing illumination in the web visualization of collada.

For what regards the issue of the center, I know that putting it to a bigger altitude it works, but the center of rotation is no more in the center of the object. Is it possible to change it?

Thanks,
Angelo

**Comment by strikerM** [comment]
Could you put these examples in a github repo, only the example code with the models?

It could be that the geometry is not triangulated or that there are more then 65536 unique indices (we use uInt16 for storing the indices).

You can rotate the model with the xRotation, yRotation and zRotation properties.
```javascript
model.xRotation = 30; //rotates 30 deg on the x axis.
 ```
There is an app example that does this:
https://github.com/NASAWorldWind/WebWorldWind/blob/develop/apps/Collada/Collada.js

**Comment by angelozinzi** [comment]
You can find the required files here https://github.com/angelozinzi/www-collada-matisse
There are two dae files: cgprova.dae is ok for visualization, pumpkin.dae is too large and is not visualized entirely.
Thanks,
Angelo

**Comment by strikerM** [comment]
I tested the models you provided.
The problem is that the geometry is indexed and there are more than 65536 unique indices.
We use uint16 for storing the indices and the index buffer overflows. 

There are a couple of things we can do to support large models:

1. Use 32 bit ints for such cases, tested with the models provided and it works.
This is straightforward to implement but consumes a bit more memory and requires the ```OES_element_index_uint``` extension.
Support for this extension seems to be very good, 95%, according to [webglstats](https://webglstats.com/webgl/extension/OES_element_index_uint)

2. Split the mesh in several meshes so that each mesh has only 65536 unique indices. 
Harder to implement and seems to be less performant since it would require extra gl calls with each split.

3. Un-index the geometry.
Harder to implement but I think it should be possible to read the geometry and duplicate each vertex as required by the index buffer.

I would go with option 1 (32 bit ints) for now.

@AkeluX @pdavidc let me know what you think.

**Comment by AkeluX** [comment]
@strikerM Dave and I discussed this issue yesterday. We concur with your assessment that 1. is the way to go about it.
In spite of the good support for this extension, we believe that there should be a check of its availability  before using it.
In addition, is there an inexpensive way to know whether the buffer will require a 32-bit index or not? This would allow us to use the extension with its possible impact on the memory only when needed.
If the extension is required, but not available, we could then print a warning in the console.
What do you think?

@angelozinzi Thank you for helping us identify this limitation. My understanding of your use case is that you want to replace the globe with an arbitrary 3D shape. In this context, I'm curious what are the advantages you see in using Web WorldWind compared to, e.g., THREE.js. Could you elaborate on this? Thank you.

**Comment by strikerM** [comment]
I agree.
We would use 32 bit ints only when necessary.
The parser will keep track of the largest index and use UInt32Array when the largest index is greater then 65535.
If the extension is required but not available we will log an appropriate message to the console.

**Comment by angelozinzi** [comment]
Dear all, thank you for the support.
To answer @AkeluX , MATISSE started as a tool for irregular objects, but now ingests data also from large objects, such as planets and moons.
Therefore we started investigating if, using World Wind Web, it will be possible to have a unique platform to visualize 3D data online: with collada for small and irregular objects and with the sphere, changing the DEM (as PlanetServer already did), for planets.

**Comment by AkeluX** [comment]
@angelozinzi Thank your for telling us more about your project. I now understand the intention. However, I have doubts about the sustainability of rendering these large Collada models instead of the globe. If you are willing to, please let us know about the conclusion of your investigation.

**Comment by angelozinzi** [comment]
Dear @AkeluX thank you for you efforts and sorry for the late reply.
However, during these months, we succeded in using a different approach to visualize high resolution shape models. In particular we used the javascript vtk.js that correctly manages also great files inside the browser.
We will probably use World Wind Web for more regular objects, such as planets (and I think this should be easier).
